qemu/tests/numa_opts.py

```diff
@@ -14,36 +14,99 @@ def run(test, params, env):
     guest.
     """
 
-    dbg("starting numa_opts test...")
+    def numa_info_guest():
+        """
+        Run 'numactl -H' in guest os and parse returned information, linux only
+
+        return: An array of (ram, cpus) tuples, where ram is the RAM size in
+                MB and cpus is a set of CPU numbers
+        """
+        session = vm.wait_for_login()
+        numa_guest_info = session.cmd_output('numactl -H')
+        re_numa_nodes = re.compile(r"([0-9]+) nodes", re.M)
+        re_numa_node_info = re.compile(r"^node ([0-9]+) (cpus|size): (.*)$", re.M)
+        nodes = re_numa_nodes.search(numa_guest_info)
+        # At least one node in vm, even no extra parameters
+        if nodes is None:
+            test.error("Couldn't get number of nodes from 'numactl -H' output")
+        nodes = int(nodes.group(1))
+        numa_guest = [[0, set()] for i in range(nodes)]
+        for nodenr, field, value in re_numa_node_info.findall(numa_guest_info):
+            nodenr = int(nodenr)
+            if nodenr > nodes:
+                test.fail("Invalid node num on 'numactl -H' output: %d", nodenr)
+            if field == 'size':
+                if not value.endswith(' MB'):
+                    test.error("Unexpected size value: %s", value)
+                megabytes = int(value[:-3])
+                numa_guest[nodenr][0] = megabytes
+            elif field == 'cpus':
+                cpus = set([int(v) for v in value.split()])
+                numa_guest[nodenr][1] = cpus
+        numa_guest = [tuple(i) for i in numa_guest]
+        # It is a known WONTFIX issue, node info of node0 and node1 is opposite
+        # with 'numactl -H' when vm have 2 nodes
+        if len(numa_guest) == 2:
+            numa_guest.reverse()
+        return numa_guest
 
     vm = env.get_vm(params["main_vm"])
+    os_type = params["os_type"]
+
+    error_context.context("starting numa_opts test...", logging.info)
 
+    # Get numa info from monitor with "info monitor"
     numa = vm.monitors[0].info_numa()
-    dbg("info numa reply: %r", numa)
+    error_context.context("info numa reply: %r" % numa, logging.info)
+
+    if os_type == 'linux':
+        # Get numa info in vm with "numactl -H",  only for Linux
+        numa_guest = numa_info_guest()
```

Comment:
You could call 'class NumaInfo()' in avocado-vt/virttest/utils_misc.py.

Reply:
In fact, none of the functions in that class can be used directly.
The only one that have relationship with this case is `get_all_node_meminfo` which get all the `meminfo`, save all the key and value...   And there is bug with it now ~
```
# cat /sys/devices/system/node/node1/meminfo 
Node 1 MemTotal:        1030376 kB
Node 1 MemFree:          276044 kB
Node 1 MemUsed:          754332 kB
Node 1 Active:           431892 kB
Node 1 Inactive:         175752 kB
Node 1 Active(anon):     301380 kB
Node 1 Inactive(anon):    10632 kB
Node 1 Active(file):     130512 kB
Node 1 Inactive(file):   165120 kB
Node 1 Unevictable:           0 kB
Node 1 Mlocked:               0 kB
Node 1 Dirty:                 0 kB
Node 1 Writeback:             0 kB
Node 1 FilePages:        307212 kB
Node 1 Mapped:           114664 kB
Node 1 AnonPages:        286264 kB
Node 1 Shmem:             11580 kB
Node 1 KernelStack:        3204 kB
Node 1 PageTables:        12288 kB
...
...
```
For these infos what we need is `MemTotal` only.  So, it does not make this case easy ~ 
Keep the current design ?

Reply:
And, I am not very sure if other cases need this information,   if needed,  move this new function to `NumaInfo` in vt is ok for me too.

Reply:
There is a function read_from_node_meminfo() to get total memory of a node. But if call this function for every node, It invokes get_all_node_meminfo() many times which is not necessary. The best solution I can think of is to add one new class variable in __init__ of class NumaInfo().
`self.node_meminfo = self.get_all_node_meminfo()` 
Then `read_from_node_meminfo() ` returns `self.node_meminfo[node_id][key]`.
And for node cpus, seem it's in another class NumaNode, get_node_cpus().
If there is a bug, I guess we should fix it. Thanks.

Reply:
Sure,  the key point is not about the existing bug, it is just a patten issue, we can fix it no matter use it or not in this case.

In my opinion, no big difference with the new added variable(`self.node_meminfo = self.get_all_node_meminfo()`) in init that you mentioned above, function `get_all_node_meminfo()` and  `read_from_node_meminfo()`.  

The main point i think is if it is worth to use it.  also for `get_node_cpus`.
There is easy way to get all the node - mem - cpu info from one command's output, why do not use it but use two class and the data need be handled too.

Reply:
IMHO, if an api is provided, we should call the api instead of implementing it again. In this case, `read_from_node_meminfo()` can return the total memory of a node, which is what exactly you need. Given that, I made the suggestion of adding new variable to avoid calling `get_all_node_meminfo()` many times.
I'm not sure about the 'easy way'. If an api is not 'easy' enough, should we ignore it or modify it?

Reply:
> IMHO, if an api is provided, we should call the api instead of implementing it again. In this case, `read_from_node_meminfo()` can return the total memory of a node, which is what exactly you need. Given that, I made the suggestion of adding new variable to avoid calling `get_all_node_meminfo()` many times.
> I'm not sure about the 'easy way'. If an api is not 'easy' enough, should we ignore it or modify it?

I understand your point, that `read_from_node_meminfo()` is easy than `get_all_node_meminfo()`, but this is not the point i care more, they do not have big difference.
The `easy way` i mean is to get mem info and cpu info in one function, not form two functions from two different class.

Reply:
IMO, involve two classes and still can not get the expect data directly is not acceptable for me.  If you insist to use the class in vt,  add a new function in vt is more reasonable in my opinion, thanks!

Reply:
I'm just trying to avoid duplicate code. There are other cases that need check node cpu inside guest too. If there is an existing api, I prefer to use it directly or update it if it's not perfect. But it's up to you. As long as the steps are right, I'm fine with it. Thanks.

Reply:
@huangyum Updated as we discussed offline, please help review again, thanks a lot!